Zend/zend_execute.c

```diff
@@ -1814,13 +1814,23 @@ static zend_always_inline void zend_fetch_dimension_address_read(zval *result, z
 			}
 		}
 	} else {
-		if (type != BP_VAR_IS && UNEXPECTED(Z_TYPE_P(container) == IS_UNDEF)) {
-			zval_undefined_cv(EX(opline)->op1.var EXECUTE_DATA_CC);
+		if (type != BP_VAR_IS) {
+			if (UNEXPECTED(Z_TYPE_P(container) == IS_UNDEF)) {
+				zval_undefined_cv(EX(opline)->op1.var EXECUTE_DATA_CC);
+			} else if (!Z_ISERROR_P(container)) {
+				zend_error(E_WARNING, "Variable of type %s does not accept array offsets", zend_get_type_by_const(Z_TYPE_P(container)));
+			}
 		}
+
 		if (/*dim_type == IS_CV &&*/ UNEXPECTED(Z_TYPE_P(dim) == IS_UNDEF)) {
 			zval_undefined_cv(EX(opline)->op2.var EXECUTE_DATA_CC);
 		}
-		ZVAL_NULL(result);
+
+		if (type != BP_VAR_IS && (dim_type == IS_CONST || ((EX(opline)+1)->op1_type == IS_VAR && EX(opline)->opcode == (EX(opline)+1)->opcode))) {
```

Comment:
This check for `dim_type == IS_CONST` looks suspicious to me. Why is it there? Wouldn't that mean that we'd yield an ERROR result for something like `$null[0]`?

On that note, I think it would be good for tests like https://github.com/php/php-src/pull/2031/files#diff-dc4f63b391c36e2dfededbde72eb4018 to var_dump the result of the lookup using `var_dump($a[0][1][2])` or similar, to check that we really get a NULL at the end, rather than an ERROR.

Reply:
I believe this is necessary for evaluating constant types from AST.   Since resolution is happening there, calls to `zend_fetch_dimension_const` set the type to IS_CONST, wherein there are no executor oplines available.  Without it the test (you also comments on) 7.phpt, will segfault.

I will go and update the tests to output the values to ensure they are not errors.

Reply:
Hm, I see. Due to a lucky coincidence the most common code-path is hardcoded to an IS_CV dim_type, which is why this does not cause issues in any of the existing tests. However, when using `var_dump(null[0])` we do end up with an UNKNOWN :/